### PR TITLE
Sync with Flathub manifest

### LINKED ISSUE #2979: Flatpak Manifest is Outdated
The develop branch is outdated compared to Flathub:

https://github.com/flathub/org.prismlauncher.PrismLauncher/blob/master/org.prismlauncher.PrismLauncher.yml

https://github.com/PrismLauncher/PrismLauncher/blob/develop/flatpak/org.prismlauncher.PrismLauncher.yml

The most serious outdated problem is that the Java Downloader is not enabled in the repo's manifest. This means that Prism Launcher is not usable as a Flatpak that is built locally, because the Flatpak cannot access the host's Java, as discussed here:

https://github.com/flathub/org.prismlauncher.PrismLauncher/issues/5

### PR DESCRIPTION
**Pull request by Arcitec**
## fix: bring back Flathub manifest improvements to the repo

- Brings back [all manifest improvements](https://github.com/flathub/org.prismlauncher.PrismLauncher/blob/master/org.prismlauncher.PrismLauncher.yml), such as the incredibly important Mojang Java Downloader (without it, the Flatpak doesn't work), and important fixes for Wayland.

Closes #2979

### CODE REVIEW

**Comment by Arcitec** [comment]
Requesting review:

@ZekeZDev @Trial97 @TheKodeToad

**Comment by Arcitec** [comment]
```sh
git clone -b flatpak-enhancements https://github.com/Arcitec/PrismLauncher.git
cd PrismLauncher
git clone https://github.com/flathub/org.prismlauncher.PrismLauncher.git flatpak-upstream
meld flatpak-upstream flatpak
```

The only thing I didn't bring over was `run-tests: true` since I assumed that keeping that on the local build was intentional.

**Review by Trial97 — APPROVED** [review]

**Review by ZekeZDev — APPROVED** [review]

**Comment by Arcitec** [comment]
Converting to draft to bring in 1 more change that TheKodeToad literally just uploaded (they committed 17 hours ago but didn't upload it until right now). It was a nice change: ~~Removing the old, external Java Flatpak runtime dependency when installing the Flatpak.~~ (Edit: This entire change was incorrect and should not have been done at Flathub either.)

https://github.com/flathub/org.prismlauncher.PrismLauncher/commits/master/org.prismlauncher.PrismLauncher.yml

**Comment by Arcitec** [comment]
Okay, this brand new Flathub change:

https://github.com/flathub/org.prismlauncher.PrismLauncher/commit/67ef3afd6abdf2ebcd3388a5300fa67d69eb44a3

Has now been brought over here:

https://github.com/PrismLauncher/PrismLauncher/compare/5215e79106f20baa8404cabcf2d8ac5a6a7190c6..ce3d43b1bac802283b6b4c113d85b32702974782

I am rebuilding locally to check that change too, but looks like a really great change which debloats the installation, since the Flatpak now uses Mojang's officially assigned Java versions.

**Comment by Arcitec** [comment]
Uh. It looks like that change actually breaks the manifest:

https://github.com/PrismLauncher/PrismLauncher/actions/runs/11502258942/job/32016814034?pr=2980

Building still works for me locally after that change, presumably because I have that `org.freedesktop.Sdk.Extension.openjdk17` and it's probably being auto-injected somehow because I have it locally already:

```
flatpak list | grep -i jdk
OpenJDK 17 SDK Extension	org.freedesktop.Sdk.Extension.openjdk17		23.08	flathub	system
OpenJDK 21 SDK Extension	org.freedesktop.Sdk.Extension.openjdk21	jdk-21.0.5+11	23.08	flathub	system
OpenJdk 8 Sdk extension	org.freedesktop.Sdk.Extension.openjdk8		23.08	flathub	system
```

I'll investigate some more. @TheKodeToad do you know what's up?

Edit: I am gonna try removing all the OpenJDKs from my system and seeing if I can still build locally.

**Comment by Arcitec** [comment]
Ah I figured it out. When building locally, it's not sandboxed, so it finds my host machine's OpenJDK "javac" at `--env=JAVA_COMPILER=/usr/lib/sdk/openjdk17/jvm/openjdk-17/bin/javac`.

The error was not caught by @TheKodeToad because Prism Launcher's Flatpak build method lacks `--sandbox` so it doesn't show what would happen during a fully "standalone, sandboxed Flatpak build without any system dependencies" (which is what Flathub uses).

I actually wanted to add `--sandbox` to my build script earlier, which would have caught this error, but I didn't do it because it would mean *having to* move the flatpak manifest from `./flatpak/` to `./` of the repo, otherwise the sandbox gets confined to `./flatpak/` (whichever directory the manifest is in) and can't find the source code. And then we'd have to mess around with a bunch of other file moves related to the Flatpak, and updating a bunch of paths in the manifest. It just didn't seem worth it even though sandboxed builds are the best way to detect issues that would happen at Flathub (which uses sandboxing when they build).

Anyway, I think it seems best that we revert that change, so that the Flatpak can build if people have no javac at all on their system. Because I see that Prism Launcher actually builds a few .java files during compilation.

As for what to do on Flathub, I suppose you should keep a close eye on the same change that was made over there.

File: flatpak/build-flatpak.sh
Comment (getchoo) [inline flatpak/build-flatpak.sh:1]: We don't need this. The build command itself is short and this makes a lot of assumptions that don't make much sense like always installing the Flatpak after it's built

File: flatpak/org.prismlauncher.PrismLauncher.yml
Comment (getchoo) [inline flatpak/org.prismlauncher.PrismLauncher.yml:6]: Why is this removed? We need Java >= 8 <= 17 to build

**Review by getchoo — CHANGES_REQUESTED** [review]

File: flatpak/org.prismlauncher.PrismLauncher.yml
Comment (Arcitec) [reply flatpak/org.prismlauncher.PrismLauncher.yml:6]: As my comments above expained, @TheKodeToad made this mistake on Flathub and will need to revert the change over there too. Because it was definitely needed by the build process. I was already in the process of reverting this change here. Hold a moment. :)

File: flatpak/build-flatpak.sh
Comment (Arcitec) [reply flatpak/build-flatpak.sh:1]: We also don't need electricity. This simplifies building tremendously and people who use the Flatpak version of the builder aren't covered at all by the wiki at the moment. Instead of navigating to `cd flatpak` and manually copying a long Wiki command every single time you need to rebuild, and manually replacing `flatpak-builder` with `flatpak run org.flatpak.Builder`, you just `./flatpak/build-flatpak.sh` instead and you're done.

It's something every developer can now remember without wasting energy and opening browsers to wikis. It's similar to the convenience of just typing `ninja` to build natively. No developer energy wasted.

The script does not need maintenance. It's gonna keep working forever. The wiki can even be improved to mention the script instead of messing with system-specific things like the existence of `flatpak-builder` natively. In fact, the `org.flatpak.Builder` (which is the officially recommended way to install Flatpak Builder) can be added as a build recommendation on the wiki, which already has a line describing how to install Flatpak dependencies.

Edit: Well I guess we do need electricity to run Prism Launcher. ;)

**Comment by Arcitec** [comment]
I have now reverted the latest change here. @TheKodeToad will need to revert the same change on the Flathub repo.

Diff from last Force Push:

https://github.com/PrismLauncher/PrismLauncher/compare/ce3d43b1bac802283b6b4c113d85b32702974782..29499f16bc153ea31c8f807a8b4f30528e44e672

Diff from start of pull request (no changes, since the removal of the Jdk was fully reverted):

https://github.com/PrismLauncher/PrismLauncher/compare/5215e79106f20baa8404cabcf2d8ac5a6a7190c6..29499f16bc153ea31c8f807a8b4f30528e44e672

As for [what an sdk-extension is in a manifest](https://docs.flatpak.org/en/latest/extension.html), the official docs say:

> "sdk-extensions can be used to install extra extensions having extension point in the parent runtime that has to be installed for the app to build. These are similarly made available during build and not in the final flatpak."

So it was only used during the Flatpak build, not afterwards during the installation. And we do need it during build. So again, TheKodeToad needs to revert the same change on Flathub, here:

https://github.com/flathub/org.prismlauncher.PrismLauncher/commit/67ef3afd6abdf2ebcd3388a5300fa67d69eb44a3

Alright, everything is restored and ready to merge again.

**Comment by Arcitec** [comment]
@Trial97 Since it seemed like you were in the process of merging, just letting you know that it's ready again and in the same state as your last review. The comments above ([starting here](https://github.com/PrismLauncher/PrismLauncher/pull/2980#issuecomment-2435556531)) explain what happened. The same fix needs to be done at the Flathub repo.

File: flatpak/build-flatpak.sh
Comment (Arcitec) [reply flatpak/build-flatpak.sh:1]: > this makes a lot of assumptions that don't make much sense like always installing the Flatpak after it's built

I missed this part. The thing is, there is no Flatpak after it's built. It's just a folder of loose data. The only thing that we *CAN* do with that folder is to install it locally as if it was a flatpak. We do not get a neat `.flatpak` file that we can distribute to anyone else.

There are some other commands that must be executed to make a `.flatpak` file.

Now, of course, I *could* add features to the build script to support doing both of those things. Adding `--install` and `--export` flags that the user can use to choose what they want to do with the build. I'll get started on that now, @Trial97.

**Comment by Arcitec** [comment]
While I was busy reverting the upstream manifest change, I only hastily read @getchoo's review. I have read it again and decided to extend the build script with features for generating `.flatpak` files for direct distribution to other computers, among other things. It will take a moment, because I just got up and am multi-tasking breakfast.

**Comment by TheKodeToad** [comment]
> I have now reverted the latest change here. @TheKodeToad will need to revert the same change on the Flathub repo.

Sorry, I am not competent with flatpak and misunderstood what getchoo told me. I am not a maintainer on the flathub repo, but somehow Scrumplex ended up approving it.

**Comment by Arcitec** [comment]
> Sorry, I am not competent with flatpak and misunderstood what getchoo told me. I am not a maintainer, but somehow Scrumplex ended up approving it.

Ah okay... Yeah don't worry about it! It's an honest and easy mistake to make. At quick glance, that line looked like it brought in Java 17 as a dependency, and of course we don't want that. But it turned out that it was just the Java Development Kit for building Prism Launcher itself, and it was not part of the final Flatpak. ;) Oh and we *need* it haha.

Anyway, as for the Flatpak build script: I am done with breakfast now and progress on the script enhancements have begun. It'll be a really neat upgrade. Edit: Almost done. Just doing builds with all options and combinations now to verify that Flatpak does the right thing. This testing will take the longest time since building this project is a bit slow, hold on. :) Although my script enables the compiler cache which really speeds it up by like 30x.

File: flatpak/build-flatpak.sh
Comment (getchoo) [reply flatpak/build-flatpak.sh:1]: > This simplifies building tremendously and people who use the Flatpak version of the builder aren't covered at all by the wiki at the moment

I think it would be better to improve the wiki (which we [are](https://github.com/PrismLauncher/prismlauncher.org/pull/597))

> manually copying a long Wiki command

This is generally the standard for our build instructions. I would say flatpak is still better off than most though; the build command boils down to `flatpak run org.flatpak.Builder --install-deps-from=flathub builddir org.prismlauncher.PrismLauncher.yml`, which is much shorter than the CMake flags [here](https://prismlauncher.org/wiki/development/build-instructions/macos/#build) for example.

Those familiar with building flatpaks should also be familiar with this command. Those who aren't will be copy-pasting the command either way, so I don't see much difference between using the direct `flatpak run org.flatpak.Builder` command and `./flatpak/build-flatpak.sh`

>  and manually replacing flatpak-builder with flatpak run org.flatpak.Builder

See the above PR

>  It's gonna keep working forever.

Until flatpak-builder changes it's CLI interface

> Now, of course, I could add features to the build script to support doing both of those things. Adding --install and --export flags that the user can use to choose what they want to do with the build. I'll get started on that now,

Which when we continue wrapping things like this, we are more at risk of. And similar to the above, these also aren't hard to add on your own when using `flatpak-builder` directly.

I think it would be much better to not abstract this, but provide snippets and point to proper [upstream documentation](https://docs.flatpak.org/en/latest/flatpak-builder.html) that is meant to explain how to use `flatpak-builder`. This would help advanced users avoid learning a new self rolled script, introduce newcomers to how `flatpak-builder` actually works (and can be applied elsewhere), and leave open all the other options of `flatpak-builder` like building without a cache, disabling tests, exporting to a repository, gpg signing, etc.

Comment (getchoo) [reply flatpak/build-flatpak.sh:1]: Could I also recommend moving this into a different PR? I'm open to discussing this further, but I would rather not have it block this PR, as it's a bit out of scope for a manifest sync, and getting these changes to development builds would be better sooner than later

Comment (Arcitec) [reply flatpak/build-flatpak.sh:1]: @getchoo Ah I see you put a ton of work into rewriting the wiki which explains why you are hesitant to automate it with a script. I'm taking a look at your changes there, and we'll see how to proceed afterwards based on what the others think.

As for "Until flatpak-builder changes it's CLI interface", that's true for the wiki too. But flatpak doesn't really change their CLI interface, so it's not a big concern.

**Comment by Arcitec** [comment]
# New Flatpak Builder for Prism Launcher

We can thank @getchoo for requesting a more flexible script. His concerns were completely valid, since he was working on a Wiki update regarding the Flatpak build process.

The new version of this Flatpak builder is incredibly flexible and **fast**, always doing *exactly* the *actually required work,* without wasting *minutes of time* on useless processes (such as never building an OSTree repo when one isn't needed). It also supports both User and System Flatpak installations.

Developers can run it with the exact combination of flags for the exact result they want. You want to generate a `.flatpak` bundle for manual distribution? You can do that. You want to install it locally? You can do that. Run it locally? You can do that.

Want to extend it later with some other, optional flags? That's easy, I included a `BUILD_EXTRA_ARGS` implementation that anyone can easily use if that ever seems necessary.

It will always take care of the *exact* developer requirements in a fast and reliable way, saving **lots** of time compared to static, manually entered build commands. It also shows every command that's being executed, to make it easy to follow what it's doing.

I have spent hours running all the different flag combinations and scenarios, with and without caches, to ensure that everything is rock-solid, including automatic installation of dependencies, system-level and user-level installations (with automatic escalation of privileges for system-installations), and every other feature.

I'll be using this for all my own projects and other projects I collaborate on too, since it's a major improvement on what I have been using there in the past.

Cleanup commands that were manually executed when deep cleaning to ensure that fresh runs of all commands work:

```sh
$ flatpak uninstall org.freedesktop.Sdk.Extension.openjdk{8,17,21} org.kde.Sdk org.prismlauncher.PrismLauncher

$ rm -rfv flatpak/.flatpak-builder flatpak/flatbuild flatpak/*.flatpak
```

Without further ado, here are the visible results of all the available commands. The logs below are abbreviated to exclude the sub-command outputs, to only focus on the important aspects.


## Help:

```sh
$ ./flatpak/build-flatpak.sh -h
Flatpak Builder for Prism Launcher.

Syntax: build-flatpak.sh [options...]

Options:
-i     Install the application after building.
-e     Export to a distributable ".flatpak" bundle.
-r     Run the application after installation (requires -i).
-u     Use the current user's Flatpak environment instead of System.
-h     Show the help.
```


## Building with default options (nothing will be installed, only useful for verifying that building the manifest works):

Note: System is the default, because that's how distros that ship pre-enabled Flathub are configured, and it's how Flathub themselves instruct people to install Flathub. If we used "user" as the default, most people would then get an error saying that "flathub" is missing. It's also very wasteful to use a user-level Flatpak environment if you have a multi-user machine, since each user would need to download everything twice, including the massive core runtimes and platforms, graphics drivers, Adwaita, Wayland, and on and on and on... which is why System *must* be the default.

```sh
$ ./flatpak/build-flatpak.sh     
Flatpak Build Options:
- Flatpak Environment: system
- Install: false
- Export: false
- Run: false

+ flatpak run org.flatpak.Builder --verbose --system --install-deps-from=flathub --force-clean --ccache flatbuild/build org.prismlauncher.PrismLauncher.yml
```


## Building with user-level Flathub:

Note: This will not work unless Flathub has been manually added to the user-level, which is a very bad idea in general (due to insane package duplication and disk space waste) and is only recommended for people who lack root privileges and cannot use the superior system-level Flatpak environment.

```sh
$ ./flatpak/build-flatpak.sh -u
Flatpak Build Options:
- Flatpak Environment: user
- Install: false
- Export: false
- Run: false

+ flatpak run org.flatpak.Builder --verbose --user --install-deps-from=flathub --force-clean --ccache flatbuild/build org.prismlauncher.PrismLauncher.yml

Installing org.kde.Sdk/x86_64/6.7 from flathub
FB: Using Flatpak version 1.14.10
FB: Running 'flatpak --user install -y --noninteractive flathub org.kde.Sdk/x86_64/6.7' on host
error: No remote refs found for ‘flathub’


$ flatpak remote-add --if-not-exists flathub --user https://dl.flathub.org/repo/flathub.flatpakrepo


$ ./flatpak/build-flatpak.sh -u
(Now it works and installs a freaking giga-million of duplicated Flatpak system-packages at the user-level instead.)
```


## Building and then installing the result:

```sh
$ ./flatpak/build-flatpak.sh -i
Flatpak Build Options:
- Flatpak Environment: system
- Install: true
- Export: false
- Run: false

+ flatpak run org.flatpak.Builder --verbose --system --install-deps-from=flathub --force-clean --ccache flatbuild/build org.prismlauncher.PrismLauncher.yml
+ flatpak build-export --verbose flatbuild/repo flatbuild/build
+ sudo flatpak install --verbose --system -y --noninteractive --reinstall ./flatbuild/repo org.prismlauncher.PrismLauncher
[sudo] password for johnny: 
(sudo is unavoidable when installing the result system-wide due to needing
 permission to introduce another Flatpak source repository, for safety reasons,
 but my runner keeps it specific to just this command)
```


## Building and then running the result:

Note: We only run via actual installations of the Prism Launcher package, to replicate the typical user's final environment perfectly, and to allow the use of Flatseal for further testing. Both flags must be explicitly provided by the developer, to avoid surprises about the fact that the Flatpak will be installed when running the development Flatpak.

```sh
$ ./flatpak/build-flatpak.sh -r
Error: The -r option also requires the -i option. Use -h to show the usage instructions.


$ ./flatpak/build-flatpak.sh -ir
Flatpak Build Options:
- Flatpak Environment: system
- Install: true
- Export: false
- Run: true

+ flatpak run org.flatpak.Builder --verbose --system --install-deps-from=flathub --force-clean --ccache flatbuild/build org.prismlauncher.PrismLauncher.yml
+ flatpak build-export --verbose flatbuild/repo flatbuild/build
+ sudo flatpak install --verbose --system -y --noninteractive --reinstall ./flatbuild/repo org.prismlauncher.PrismLauncher
+ flatpak run --system org.prismlauncher.PrismLauncher
```


## Building and exporting as a distributable Flatpak file:

Note: The resulting `.flatpak` file can be copied to other computers and installed via `flatpak install ./org.prismlauncher.PrismLauncher.flatpak` without needing any build environment or dependencies.

```sh
$ ./flatpak/build-flatpak.sh -e
Flatpak Build Options:
- Flatpak Environment: system
- Install: false
- Export: true
- Run: false

+ flatpak run org.flatpak.Builder --verbose --system --install-deps-from=flathub --force-clean --ccache flatbuild/build org.prismlauncher.PrismLauncher.yml
+ flatpak build-export --verbose flatbuild/repo flatbuild/build
+ flatpak build-bundle --verbose --runtime-repo=https://flathub.org/repo/flathub.flatpakrepo flatbuild/repo org.prismlauncher.PrismLauncher.flatpak org.prismlauncher.PrismLauncher


$ ls -al flatpak/*.flatpak
-rw-r--r--. 1 johnny johnny 168340096 Oct 24 21:58 flatpak/org.prismlauncher.PrismLauncher.flatpak
```


## Combining all options, just because we can:

Note: This will build, then install the result, then export as a `.flatpak`, and finally run the installed application.

```sh
$ ./flatpak/build-flatpak.sh -ier
Flatpak Build Options:
- Flatpak Environment: system
- Install: true
- Export: true
- Run: true

+ flatpak run org.flatpak.Builder --verbose --system --install-deps-from=flathub --force-clean --ccache flatbuild/build org.prismlauncher.PrismLauncher.yml
+ flatpak build-export --verbose flatbuild/repo flatbuild/build
+ sudo flatpak install --verbose --system -y --noninteractive --reinstall ./flatbuild/repo org.prismlauncher.PrismLauncher
+ flatpak build-bundle --verbose --runtime-repo=https://flathub.org/repo/flathub.flatpakrepo flatbuild/repo org.prismlauncher.PrismLauncher.flatpak org.prismlauncher.PrismLauncher
+ flatpak run --system org.prismlauncher.PrismLauncher


$ ls -al flatpak/*.flatpak
-rw-r--r--. 1 johnny johnny 168342496 Oct 24 22:04 flatpak/org.prismlauncher.PrismLauncher.flatpak
```

**Comment by Arcitec** [comment]
Now it's time for a well-deserved Prism Launcher Minecraft break. :sweat_smile:

**Review by Ryex — APPROVED** [review]
I initially had concerns that I didn't voice (was too tired to write them all down last night)
I'd prefer to leave the final approval to getchoo but things look good now.

**Review by Trial97 — APPROVED** [review]

**Comment by Arcitec** [comment]
@Ryex Thanks, I appreciate it! The initial variant was basically "take some of the code that I use on other projects, cut it down to the essentials and replicate Prism Launcher's (old) Wiki instructions". But I am really glad to have done this rewrite because it will  not just help Prism Launcher, it will also greatly benefit my other projects and others that I collaborate on as well. It's also easier to use than static Wiki instructions due to the dynamic nature of what needs to be done for various desired outcomes. All in all, it was worth the extra work.

**Review by getchoo — CHANGES_REQUESTED** [review]
> We can thank @getchoo for requesting a more flexible script.

This *really* isn't what I meant by my previous comments, and I'm sorry that it came across that way. I brought up the inflexibilities with script as an example of why we *shouldn't* be making it

I think the amount of documentation we have in the above comment is a good example of this. It is *extremely* similar to existing upstream documentation such as the base [Flatpak Builder page](https://docs.flatpak.org/en/latest/flatpak-builder.html), ["Building your first Flatpak"](https://docs.flatpak.org/en/latest/first-build.html), and the [command reference](https://docs.flatpak.org/en/latest/flatpak-builder-command-reference.html). There is no reason we should be putting on our spin on things to this extent when the base command isn't all that complicated, but offers so much other functionality people would realistically use -- and while yes, you have added more very specific features, there are still many things missing like using repos besides Flathub, setting OSTree branch names, building and installing in one go, or even just passing arbitrary arguments to the commands here like `--jobs`. **Keep in mind this is not me saying to add these functions**, but just a short list of examples of *why* we shouldn't abstract this; we shouldn't need to add explicit functionality already existing in `flatpak-builder` to this script whenever we need it. `flatpak-builder` itself is more than good enough

Again, I would recommend opening a separate PR with this script. It is now even further out of scope from a manifest sync, and should have it's own review process

**Comment by Arcitec** [comment]
@getchoo I'll separate it to its own pull request, if you're sure about that, absolutely. :)

I do not agree with the comment that we "may want to do even more different things" though. This handles **all** of the realistic developer requirements for Prism Launcher: Fast local builds, fast local testing, fast bundle export for manual copying to other computers. Without any scope creep turning it into a million-feature wrapper.

Your examples are perfect examples of where someone can easily see what commands are being executed by the script and tweak them manually, and a developer can even add a flag which extends `BUILD_EXTRA_ARGS` with anything you might desire in the future.

The Wiki PR is extremely inflexible. It forces `--user` in its examples which does not work for typical distros which use System-level Flatpak. It exports to a `--repo` on every build, which wastes minutes of time.

While you _could_ write a few hundred paragraphs on the Wiki documenting all the variants of every command on the wiki, which all have to be manually copy-pasted during builds, it would also be possible to save everyone's time and let them wrap the process in a fast way that doesn't do any slow, useless extra work.

The existence of a wrapper does not *prevent* someone from doing anything differently if they truly want to. It just ensures that people don't have to scour a wiki, troubleshoot all the commands, remove the suboptimal arguments such as `--repo`, remove `--user`, add `--system`, manually add `sudo` (which does not work with `flatpak-builder` by the way, since compilation refuses to run as root!) and so on, and to then repeat the entire process every single time they want to build.

The script handles all of the normal developer scenarios for Prism Launcher and lowers the barrier to entry for new contributors. It just works.

Should I move it to a new pull request?

**Comment by Trial97** [comment]
> Should I move it to a new pull request?

Yes please(it would be easier to review that way)

**Comment by Arcitec** [comment]
Alright gimme a minute.

**Comment by Arcitec** [comment]
This pull request is now solely for the manifest.

The separate pull request is up at #2988 when you have a bit of spare time to take a look.

**Review by getchoo — APPROVED** [review]
Thanks!

**Comment by Arcitec** [comment]
Excellent, glad to see it merged! And apologies for the original scope creep in this PR! :)

**Pull request merged**
